### PR TITLE
Prevent nil panic from derefeence nil context in authorization middleware

### PR DESCRIPTION
**Pull request by troydai**
**Pull Request check list**

- [x] Commit conforms to CONTRIBUTING.md?
- [x] Proper tests/regressions included?
- [x] Documentation updated?

**Affected functionality**

Prevent a nil panic when a database operation failure happens during retrieving entries.

**Description of change**

Nil panic happens in authorization middleware's Preprocess function when a nil context is passed down from a failed database operation while fetching entries in function WithCallerEntries. This change addresses this risk by introduce a new variable to hold the returned context so as to avoid dereferencing a potential nil ctx.

Extended explanation:

WithCallerEntries returns error when a database operation failure happens. The error itself is not fatal, however, it returns a nil context. Returns zero value when errors is correct behavior. However, the reconcileResult function and opaAuth function bubble up the nil context as if it is populated. In the authorization middleware's Preprocess function, it attempts to retrieve a logger from the context thus cased a nil panic.

Fix:

The fix is simple: Since the dereference happens for the purpose of fetching a logger, a new variable is introduced to store returned context and deferencing will happen on the original contexty. This is simplest fix of all the forms. Other alternative fix include introduce a separate variable to hold mutated context and I believe it overcomplicates the code.

Other changes:

This change also include a fix in the `authorization_opa.go` to explicitly returns nil context in when functions errored out. It doesn't change the logic because it is already returning nil.

**Which issue this PR fixes**

### CODE REVIEW

File: pkg/server/api/middleware/authorization.go
Comment (azdagron) [inline pkg/server/api/middleware/authorization.go:61]: Can we also not alias the ctx variable? I think that would be a little more robust towards future refactors.

```
authedCtx, allow, err := m.opaAuth(ctx, req, methodName)
....
if allow {
    return authedCtx, nil
}
```

**Review by azdagron — COMMENTED** [review]
Thank you, @troydai. Excellent catch. Just one small suggestion...

**Review by azdagron — APPROVED** [review]
\o/

**Comment by troydai** [comment]
Thank you. Updated the PR description, will merge once all the checks pass.

**Comment by azdagron** [comment]
> will merge once all the checks pass

Only maintainers have permissions to merge. We'll take care of it :)

**Comment by troydai** [comment]
Got it. There is a race condition test failed. Glance at it, it doesn't seem to be related to this change. Shall I dig into it or there will be a retry?

**Comment by azdagron** [comment]
It's a racy test. I'll retrigger it after the rest have finished.

**Pull request merged**
